### PR TITLE
Improve performance and add filtering to ValueSet based code selection.

### PR DESCRIPTION
**Pull request by kaicode**
Following on from [709](/synthetichealth/synthea/pull/709) where John Grimes added support for code selection using FHIR ValueSets..

This pull request: 
- Improves the performance of generating large sets of patients by caching the first 1K codes for each ValueSet and not repeating the FHIR API fetch. Codes are randomly selected from this set as before. 
- Adds the ability to use an implicit value set with a term filter. This is in the FHIR spec but was not supported by the previous implementation.

Work was completed by [Elimu](elimu.io) on behalf of SNOMED International.

Another pull request will be raised against the module builder to allow ValueSets to be specified in the user interface.

### CODE REVIEW

**Comment by jawalonoski** [comment]
Thank you for the pull request! Performance improvements are always welcome.

It may take some time to review, so thank you in advance for your patience.

File: build.gradle [hunk 1]
```diff
@@ -53,11 +53,14 @@ dependencies {
...
   compile group: 'org.apache.commons', name: 'commons-math3', version: '3.6.1'
   compile group: 'org.apache.commons', name: 'commons-text', version: '1.2'
+  compile group: 'commons-validator', name: 'commons-validator', version: '1.4.0'
 
   compile group: 'org.opencds.cqf', name: 'cql-engine', version: '1.3.12'
   compile group: 'info.cqframework', name: 'cql', version: '1.3.17'
   compile group: 'info.cqframework', name: 'model', version: '1.3.17'
   compile group: 'info.cqframework', name: 'cql-to-elm', version: '1.3.17'
+  
+  compile group: 'org.springframework', name: 'spring-web', version: '3.0.2.RELEASE'
```
Comment (eedrummer) [inline build.gradle:63]: Is there a reason to use version 3.0.2.RELEASE? Based on https://mvnrepository.com/artifact/org.springframework/spring-web, this is a release from April 2010. I'm hesitant on adding a new dependency that is two major releases behind the current version.

File: src/main/java/org/mitre/synthea/export/CSVExporter.java [hunk 2]
```diff
@@ -383,7 +385,16 @@ public void export(Person person, long time) throws IOException {
       String payerID = encounter.claim.payer.uuid;
 
       for (HealthRecord.Entry condition : encounter.conditions) {
-        condition(personID, encounterID, condition);
+        /* condition to ignore codes other then retrieved from terminology url */
+        if (!StringUtils.isEmpty(Config.get("generate.terminology_service_url"))
+            && !RandomCodeGenerator.selectedCodes.isEmpty()) {
+          if (RandomCodeGenerator.selectedCodes.stream().filter(code -> code.code.equals(condition.codes.get(0).code))
```
Comment (eedrummer) [inline src/main/java/org/mitre/synthea/export/CSVExporter.java:391]: Minor checkstyle violation here

File: src/main/java/org/mitre/synthea/helpers/RandomCodeGenerator.java [hunk 3]
```diff
@@ -1,191 +1,115 @@
...
 import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestTemplate;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
```
Comment (eedrummer) [inline src/main/java/org/mitre/synthea/helpers/RandomCodeGenerator.java:25]: There are a few checkstyle violations in this file

**Comment by kaicode** [comment]
Thank you for your review @eedrummer. 

@amol-wani-elimu Would you be able to fix the style issues and set a more recent [Spring Web release](https://search.maven.org/artifact/org.springframework/spring-web) please?

**Comment by amol-wani-elimu** [comment]
> Thank you for your review @eedrummer.
> 
> @amol-wani-elimu Would you be able to fix the style issues and set a more recent [Spring Web release](https://search.maven.org/artifact/org.springframework/spring-web) please?

Yes @kaicode, yes I have started working on comments, will push once done with fixes.

**Comment by eedrummer** [comment]
@amol-wani-elimu Thanks for making the changes. The `master` branch has been updated since you have issued this PR and now there is a conflict. Could you resolve this? After that, I think we are good to go.

**Comment by kaicode** [comment]
@eedrummer rebased against `master`. Only conflict was an import statement and tests passing so looks good to me.

**Pull request merged**